src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs

```diff
@@ -15,71 +16,56 @@ internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
     {
         public static readonly Checksum Null = new Checksum(Array.Empty<byte>());
 
-        private readonly byte[] _checkSum;
-        private int _lazyHash;
+        private Sha1Hash _checkSum;
 
-        public Checksum(byte[] checksum)
+        public unsafe Checksum(byte[] checksum)
         {
-            // 0 means it is not initialized
-            _lazyHash = 0;
-            _checkSum = checksum;
-        }
-
-        public bool Equals(Checksum other)
-        {
-            if (other == null)
+            if (checksum.Length == 0)
```

Comment:
It might be nice to have two constructors, only one of which can be used inside the class and which allows or an empty array.  that way no one outside accidentally does this.